### PR TITLE
[X86][FP16][BF16] Improve vectorization of fcmp

### PR DESCRIPTION
**Pull request by phoebewang**

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -23333,12 +23339,8 @@ static unsigned translateX86FSETCC(ISD::CondCode SetCCOpcode, SDValue &Op0,
 
 /// Break a VSETCC 256-bit integer VSETCC into two new 128 ones and then
 /// concatenate the result back.
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:23341]: `VSETCC 256-bit VSETCC`

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -23333,12 +23339,8 @@ static unsigned translateX86FSETCC(ISD::CondCode SetCCOpcode, SDValue &Op0,
...
-static SDValue splitIntVSETCC(EVT VT, SDValue LHS, SDValue RHS,
-                              ISD::CondCode Cond, SelectionDAG &DAG,
-                              const SDLoc &dl) {
-  assert(VT.isInteger() && VT == LHS.getValueType() &&
-         VT == RHS.getValueType() && "Unsupported VTs!");
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:23340]: Replace with this?
`assert(VT.isInteger() && LHS.getValueType() == RHS.getValueType() && "Unsupported VTs!");`

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 3]
```diff
@@ -23483,14 +23485,40 @@ static SDValue LowerVSETCC(SDValue Op, const X86Subtarget &Subtarget,
   SDLoc dl(Op);
 
   if (isFP) {
-    MVT EltVT = Op0.getSimpleValueType().getVectorElementType();
+    MVT SVT = Op0.getSimpleValueType();
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:23488]: We usually use SVT for scalar VTs - maybe OpVT? The getSimpleValueType can also be merged with the isFP check above

**Review by RKSimon — COMMENTED** [review]
A few minors

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 1 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:23341]: Done

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 2 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:23340]: Done

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by rnk** [comment]
This appears to have caused an regression on our internal codebase, generating this error message:
```
fatal error: error in backend: Cannot select: 0x13ef7f4be070: v16f32 = X86ISD::CMPP 0x13ef7f4bd2a0, 0x13ef7f5343f0, TargetConstant:i8<7>
  0x13ef7f4bd2a0: v16f32 = X86ISD::CVTPH2PS 0x13ef7d2eb0e0
    0x13ef7d2eb0e0: v16i16 = bitcast 0x13ef7f4bec40
      0x13ef7f4bec40: v16f16,ch = load<(load (s256) from %ir.142, align 1)> 0x13ef6dcf52a0, 0x13ef7f4bd230, undef:i64
        0x13ef7f4bd230: i64 = add 0x13ef7f534770, Constant:i64<96>
          0x13ef7f534770: i64 = add 0x13ef7d2eb3f0, 0x13ef7f4be540
            0x13ef7d2eb3f0: i64,ch = CopyFromReg 0x13ef6dd3d270, Register:i64 %1
              0x13ef6dcf5e70: i64 = Register %1
            0x13ef7f4be540: i64 = shl 0x13ef7f534a80, Constant:i8<1>
              0x13ef7f534a80: i64,ch = CopyFromReg 0x13ef6dd3d270, Register:i64 %4
                0x13ef7f4bd700: i64 = Register %4
              0x13ef7f4bd850: i8 = Constant<1>
          0x13ef6dcf58c0: i64 = Constant<96>
        0x13ef646602a0: i64 = undef
  0x13ef7f5343f0: v16f32 = BUILD_VECTOR ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>, ConstantFP:f32<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
    0x13ef7f4bd1c0: f32 = ConstantFP<0.000000e+00>
  0x13ef6dcf54d0: i8 = TargetConstant<7>
```

I will start a reduction, but I wanted to give you a heads up that we may need to revert or fix forward quickly.

**Comment by phoebewang** [comment]
Thanks for reporting this. I'll try a quick fix if there's a reproducer. I can help with reduction too :)

**Comment by bgra8** [comment]
@phoebewang here's the reproducer for the crash:

repro.ll:
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-generic-linux-gnu"

define void @_test_func(<16 x half> %0) #0 {
  %2 = fcmp ord <16 x half> %0, zeroinitializer
  %3 = sext <16 x i1> %2 to <16 x i32>
  %4 = shufflevector <16 x i32> %3, <16 x i32> zeroinitializer, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
  %5 = tail call <8 x i16> @llvm.x86.sse2.packssdw.128(<4 x i32> %4, <4 x i32> zeroinitializer)
  %6 = shufflevector <8 x i16> %5, <8 x i16> zeroinitializer, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15>
  %7 = bitcast <16 x i16> %6 to <32 x i8>
  store <32 x i8> %7, ptr null, align 1
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(none)
declare <8 x i16> @llvm.x86.sse2.packssdw.128(<4 x i32>, <4 x i32>) #1

attributes #0 = { "target-features"="+aes,+avx,+avx2,+avx512f,+avx512vnni,+cmov,+crc32,+cx16,+cx8,+evex512,+f16c,+fma,+fxsr,+mmx,+pclmul,+popcnt,+prfchw,+sse,+sse2,+sse3,+sse4.1,+sse4.2,+ssse3,+x87,+xsave" }
attributes #1 = { nocallback nofree nosync nounwind willreturn memory(none) }
```

Repro command:

```
$ llc < repro.ll
```

This crashes at this revision and compiles at the previous one.

**Comment by phoebewang** [comment]
Thanks for the repro @bgra8! I think there might be some problem with compiler options in you case. With KNL support been deprecated, there's no target that supports AVX512F but not AVX512VL. Anyway, https://github.com/llvm/llvm-project/pull/119309 would be able to solve the crash issue.

**Comment by bgra8** [comment]
@phoebewang adding or replacing `-mavx512vl` in the compilation command doesn't prevent the crash.

**Comment by phoebewang** [comment]
> @phoebewang adding or replacing `-mavx512vl` in the compilation command doesn't prevent the crash.

No sure what's the problem in the original case, but the IR works with `avx512vl` https://godbolt.org/z/q6YnfYsK1

Please attach another repro if the fix doesn't work for you.

**Comment by bgra8** [comment]
@phoebewang your patch in https://github.com/llvm/llvm-project/pull/119309 fixes our original issue! Thanks a lot!

**Pull request merged**